cpp/src/cider/exec/processor/DefaultBatchProcessor.cpp

```diff
@@ -55,25 +55,27 @@ DefaultBatchProcessor::DefaultBatchProcessor(
     const BatchProcessorContextPtr& context,
     const cider::exec::nextgen::context::CodegenOptions& codegen_options)
     : context_(context) {
-  auto allocator = context->getAllocator();
-  if (plan->hasJoinRel()) {
-    // TODO: currently we can't distinguish the joinRel is either a hashJoin rel
-    // or a mergeJoin rel, just hard-code as HashJoinHandler for now and will refactor to
-    // initialize joinHandler accordingly once the
-    joinHandler_ = std::make_shared<HashProbeHandler>(shared_from_this());
-    this->state_ = BatchProcessorState::kWaiting;
-  } else if (plan->hasCrossRel()) {
-    joinHandler_ = std::make_shared<CrossProbeHandler>(shared_from_this());
-    this->state_ = BatchProcessorState::kWaiting;
-  }
-
-  auto translator =
-      std::make_shared<generator::SubstraitToRelAlgExecutionUnit>(plan->getPlan());
-  RelAlgExecutionUnit ra_exe_unit = translator->createRelAlgExecutionUnit();
+  initJoinHandler(plan);
+  auto translator = generator::SubstraitToRelAlgExecutionUnit(plan->getPlan());
+  RelAlgExecutionUnit ra_exe_unit = translator.createRelAlgExecutionUnit();
```

Comment:
seems create the RelAlgExecutionUnit twice here?

Reply:
just remove shared_ptr: delete old, add new

Reply:
I can't quite get it. Seems it will do the EU generation again.

Reply:
discussion offline: we have two APIs now, this is the old API that complies from substrait plan directly.
I'll add some comments in code